PROTOCOL.md

```diff
@@ -564,22 +565,63 @@ The following is an example `rowIdHighWaterMark` action:
 }
 ```
 
+### Domain Metadata
+The domain metadata action contains a configuration (string-string map) for a named metadata domain. Two overlapping transactions conflict if they both contain a domain metadata action for the same metadata domain.
+
+There are two types of metadata domains:
+1. **User-controlled metadata domains** have names that start with anything other than the `delta.` prefix. Any Delta client implementation or user application can modify these metadata domains, and can allow users to modify them arbitrarily. Delta clients and user applications are encouraged to use a naming convention designed to avoid conflicts with other clients' or users' metadata domains (e.g. `com.databricks.*` or `org.apache.*`).
+2. **System-controlled metadata domains** have names that start with the `delta.` prefix. This prefix is reserved for metadata domains mentioned in the Delta spec (i.e. as part of some table feature), and Delta client implementations must not allow users to modify the metadata for system-controlled domains. A Delta client implementation should only update metadata for system-controlled domains that it knows about and understands. System-controlled metadata domains are used by various table features and each table feature may impose additional semantics on the metadata domains it uses.
```

Comment:
@ryan-johnson-databricks , somehow I lost your previous comment in my git comment history. Let me try to understand what you mean here.

>  I could imagine system-controlled domains for which the default domain handling suffices (writers don't erase or modify its content), but which are documented in the Delta spec and thus can be manipulated by a client who actually understands them?

The current spec doesn't mention how the users who understand a domain manipulate the domain metadata and we should say something here. Is that right understanding?
I don't have an answer here honestly since it is depending on the table feature who used this domain.  I am thinking any future table feature who utilizes the domain metadata should specify it in their own delta spec section. But any suggestions here is appreciated.

> This would cover any kind of auxiliary/optional information that isn't important enough to force the compat requirements that come with a table feature, but where it would still be nice if other readers didn't actively destroy the domain's info? Maybe something like query optimizer stats that don't need to be perfectly up to date, so it's better to let an older/unaware client to propagate the existing (now stale) entry, rather than block that client or delete the stats?

sorry, I don't follow here. Is this something related how to handle two concurrent transitions that commit the same domain metadata?

Reply:
You are correct that a table feature should specify how to interpret, modify, and resolve conflicts for a given system-controlled metadata domain. I was referring to the possibility of two sub-flavors:
1. A writer table feature might include some metadata domains that aware readers _could_ leverage in some way to their advantage, but which unaware readers can safely ignore without causing correctness issues.
2. A "ghostly" table feature which both readers and writers can safely ignore if they don't know about it. Such features should still be documented in the Delta spec, but not as table features that might cause compat concerns -- clients can choose whether to implement support or not.

The hypothetical optimizer stats example I gave above would be an example of a "ghost" feature -- if a reader ignores such stats, they might produce an inferior query plan but otherwise no harm. If a writer ignores such stats (as in, merely propagates them blindly per the metadata domain spec, but does not try to update them in response to changing data), then the stats would slowly go stale and become less useful over time. But that's actually not a problem -- most database engines I know of anyway don't try to recompute such stats at every update, because it's too expensive and the stats normally evolve quite slowly.

> Is this something related to how to handle two concurrent transitions that commit the same domain metadata?

An unaware writer should never try to update an unrecognized metadata domain, so it should never cause a transaction conflict -- "blindly propagating" the metadata domain means the commit should not mention it (neither as added nor as removed), and thus satisfies the general conflict resolution rules for metadata domains which this PR already includes.

Reply:
I think it's enough not _require_ that every system-controlled metadata domain _must_ be part of a table feature (it just needs to be documented in the Delta spec). That should suffice, when combined with the proposed wording on reader/writer requirements: If a system-controlled metadata domain needs special handling, that's when it becomes a breaking change and needs to be part of a table feature (just like any other breaking change)

Reply:
Got it, I removed the `(i.e. as part of some table feature)` in this sentence without requiring that it must be in a table feature. Thanks for the feedback!